### ISSUE TITLE
Proposed YAML syntax change to allow hydration free energies

### ISSUE THREAD

**Issue by andrrizzi** [choderalab/yank#416]
Here's a proposed syntax which involves specifying systems for each phase and moving protocols descriptions inside the `experiments` section (or this section would end up with only one subkey)

```yaml
molecules:
  Abl:
    ...
  imatinib
    ...
  bosutinib
    ...

solvents:
  PME:
    ...
  vacuum:
    nonbonded_method: NoCutoff

systems:
  complex:  # for complex phase of binding free energies
    receptor: Abl
    ligand: !Combinatorial [imatinib, bosutinib]
    solvent: PME
  solvent:  # for solvent phase of binding free energies, and first phase of hydration free energy
    ligand: !Combinatorial [imatinib, bosutinib]
    solvent: PME
  vacuum:  # for second phase of hydration free energy
    ligand: !Combinatorial [imatinib, bosutinib]
    solvent: vacuum

experiments:  # an example of hydration free energy calculation
  solvent:  # this is an arbitrary name for the phase
    cycle_direction: +
    lambda_electrostatics: [1.00, 0.95, 0.90, 0.85, 0.80, ...]
    lambda_sterics: [1.00, 1.00, 1.00, 1.00, 1.00, ...]
  vacuum:  # this is an arbitrary name for the phase
    cycle_direction: -
    lambda_electrostatics: [1.00, 0.95, 0.90, 0.85, 0.80, ...]
    lambda_sterics: [1.00, 1.00, 1.00, 1.00, 1.00, ...]
  phases_systems: [solvent, vacuum]  # first one is for + phase, second for - phase
  options:
    ...  # experiment-specific options
```
Remember that sequences of experiments are already supported, so this could be a valid syntax
```yaml
binding_experiment:
  complex:  # this is an arbitrary name for the phase
    cycle_direction: +
    lambda_electrostatics: [1.00, 0.95, 0.90, 0.85, 0.80, ...]
    lambda_sterics: [1.00, 1.00, 1.00, 1.00, 1.00, ...]
  solvent:  # this is an arbitrary name for the phase
    cycle_direction: -
    lambda_electrostatics: [1.00, 0.95, 0.90, 0.85, 0.80, ...]
    lambda_sterics: [1.00, 1.00, 1.00, 1.00, 1.00, ...]
  phases_systems: [complex, solvent]

hydration_experiment:
  ...  # exactly like before

experiments: [binding_experiment, hydration_experiment]
```
also it would be possible to specify combinations like
```yaml
experiments:  # an example of hydration free energy calculation
  solvent:
   ...  # as before
  vacuum:
   ...  # as before
  phases_systems: !Combinatorial [[solvent1, vacuum1], [solvent2, vacuum2]]
```
I think this could be pretty flexible. Any use case that I'm missing?

**Comment by andrrizzi** [comment]
Ah! I just remembered why we went with a single system specification for both phases. The first example above generates 4 experiments, mixing phases with imatinib and bosutinib!

I think we are stuck with a single system specification for both phases if we want to do combinations of molecules. At this point the **type of calculation can be encoded in the system**. For example:
```yaml
systems:
  binding_system_example:  # receptor, ligand, solvent specify binding free energy calculation
    receptor: Abl
    ligand: !Combinatorial [imatinib, bosutinib]
    solvent: PME
  hydration_system_example:  # molecule, solvent specify solvation
    molecule: imatinib
    solvent: PME
  partition_system_example:
    molecule: imatinib
    solvent1: RF-tip3p
    solvent2: RF-octanol
```
Better solutions?

**Comment by jchodera** [comment]
On the last proposal, how do we tell what kind of simulation we are
planning to run? Is it
* If receptor and ligand are specified, run a binding free energy
calculations
* If a molecule is specified (maybe rename this to 'solute'?) run a
hydration free energy calculation
* If a solute and solvent1 and solvent2 are specified, run partition
coefficient calculation (maybe can generalize to more solvents too?)

This sounds really convenient, though may run into extensibility problems
later. Since the key here is how the specified system components are
combined to make real systems, adding a key that specifies which system
combining function to use (binding, hydration, partition) may be one way to
generalize this. The appropriate combining function would then be in charge
of checking for the components it wants to use.

I'm still a bit worried about specifying both the long range electrostatics
treatment (OBC, PME, RF) and the solvent model (TIP3P) in the same key,
since this seems a bit inflexible in the future. If there was a simple way
we could split this into a nonbonded method, optional GB model, and solvent
model, that would be awesome. But I understand if things are too complex to
do this right now.

**Comment by andrrizzi** [comment]
> If receptor and ligand are specified, run a binding free energy calculations

Yes, that's the idea.

> adding a key that specifies which system combining function to use (binding, hydration, partition) may be one way to generalize this

The first design I thought had this, but the components of the system will have to be different anyway (`receptor` `ligand` `solvent` for absolute binding, `molecule` `solvent` for solvation, etc.), so **I thought that key would be only redundant information, which means I'd have to add another step during setup to check if that key and the components are consistent with each other**. Do you have a concrete example of some type of calculation where this design might be not extensible?

> I'm still a bit worried about specifying both the long range electrostatics treatment (OBC, PME, RF) and the solvent model (TIP3P) in the same key

**In the example above, `RF-tip3p` is only an ID referring to a solvent specified above** with the usual syntax, where those three components are already split.

**Comment by jchodera** [comment]
> Do you have a concrete example of some type of calculation where this design might be not extensible?

I'm just thinking ahead, so this shouldn't really constrain us *too* much, but being able to be explicit about how we set things up might be very helpful:
```YAML
systems:
  # ligand:receptor binding in explicit solvent
  explicit_solvent_binding_in_solution:
    scheme: binding_explicit
    receptor: Abl
    ligand: !Combinatorial [imatinib, bosutinib]
    solvent: tip4p-ew
    nonbondedMethod: PME
  # ligand:receptor binding in implicit solvent
  implicit_solvent_binding_in_solution:
    scheme: binding_implicit
    receptor: Abl
    ligand: !Combinatorial [imatinib, bosutinib]
    gbsa: OBC1
    nonbondedMethod: NoCutoff
  # ligand:receptor binding in membrane
  explicit_solvent_binding_in_membrane:
    scheme: binding_explicit_membrane
    receptor: beta2AR
    ligand: !Combinatorial [salmeterol, albuterol]
    solvent: tip3p
    nonbondedMethod: PME
  # hydration free energy in explicit solvent
  hydration_free_energy_explicit:
    scheme: hydration_explicit
    solute: !Combinatorial [methanol, butanol, ethanol]
    solvent: tip3p
    nonbondedMethod: CutoffPeriodic
  # hydration free energy in implicit solvent
  hydration_free_energy_implicit:
    scheme: hydration_implicit
    solute: !Combinatorial [methanol, butanol, ethanol]
    gbsa: OBC1
    nonbondedMethod: NoCutoff
  # partition coefficient among two or more solvents
  partition_coefficient:
    scheme: partition_coefficient # this setup scheme knows to generate GAFF solvents
    solute: !Combinatorial [methanol, butanol, ethanol]
    solvent1: tip3p
    solvent2: cyclohexane
    solvent3: octanol
    nonbondedMethod: PME
  # solubility
  solubility:
    scheme: amorphous_solubility # this scheme knows how to generate amorphous solids for the solid phase
    solute: !Combinatorial [methanol, butanol, ethanol]
    solvent: tip3p
    nonbondedMethod: PME
```


  hydration_system_example:  # molecule, solvent specify solvation
    molecule: imatinib
    solvent: PME
  partition_system_example:
    molecule: imatinib
    solvent1: RF-tip3p
    solvent2: RF-octanol
```

**Comment by andrrizzi** [comment]
Addressed by #420 .
